text/0219-ecs-service-extensions.md

```diff
@@ -0,0 +1,503 @@
+to create their own extension: `MyServiceCustomizer` which modifies the underlying L2
+construct properties directly.
+
+Another drawback of this approach is that it creates a dependence on the L3 construct.
+This `Service` class implements attractive new features that can not be easily implemented
+at L2 without deep knowledge of the feature itself.
+via these L3 construct extensions it will be a massive effort to reimplement the same
+results directly at the L2 level. Basically once you start implementing your service
+using this L3 construct it would be almost impossible to refactor your CDK app back into
+lower level L2 constructs because the surface area and logic that the L3 construct
+implements is very broad.
+
+1. __Do we anticipate certain extensions conflicting with each other?__
+
+   Yes it is possible that extensions may conflict with each other. It is the responsibility
+   of each extension to check for other extensions at the time of being added to the service
+   and throw an exception if there is an expected conflict.
+
+2. __Given that ECS patterns are meant to push best practices in application
+   infrastructure/architecture, how will we provide that guidance with extensions?__
+
+   The goal of these extensions is to do the right thing by default, similar to the
+   existing ECS patterns, but the user just has more choices about what combinations
+   of feature integrations they want to enable. When using the built-in extensions there
+   should be no "bad" choices. If a user wants to leave out an extension it is safe to
+   do so, and if they chose to enable it there should be no gotchas from adding another
+   extension. The goal is to make this hard to misuse, but still powerful and extendable.
+
+# Rationale and Alternatives
+
+One alternative approach to this is to try to enhance the lower level L2 methods with
+an extension type system. The existing
+[`TaskDefinition.addExtension()`](https://docs.aws.amazon.com/cdk/api/latest/docs/@aws-cdk_aws-ecs.TaskDefinition.html#add-wbr-extensionextension)
+method is the beginning of an attempt at this.
+
+However attempting to build out extensions on the L2 level will fail to fully meet customer
+expectations because the features they want to enable bridge multiple different
+types of L2 construct. For example in order for an extension like App Mesh to work
+properly it must change the service to have a cloud map configuration, change the
+task definition IAM role to have the right level of access, change the task definition
+to add proxy settings, attach a container to the task definition, and create and link
+App Mesh virtual service and virtual node to the service. Because the full feature
+requires touching all of these different constructs and resulting resources it is not
+possible to fully implement an extension system at the L2 level. For this reason the
+existing `TaskDefinition.addExtension()` is not a viable alternative.
```

Comment:
Not strictly true. Yes, the interface of the `IExtension` itself should be narrow, but there's no need for the **implementation** to live in the L2 library that would restrict its dependencies. 

The specific extensions themselves could totally be in a different library and reach out to AppMesh and whatnot.